### PR TITLE
[FLINK-4686] [table] Add possibility to get column names

### PR DESCRIPTION
**Pull request by wuchong**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.
- [ x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)
- [x ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added
- [x ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

This PR introduces `Schema` to allow us get column names and types.

``` java
Schema s = table.schema();
TypeInformation<?> type = s.getType(1);
TypeInformation<?> type = s.getType("col");
String s = s.getColumnName(1);
String[] s = s.getColumnNames();
```

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/Schema.scala [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+ */
+package org.apache.flink.api.table
+
+import org.apache.flink.api.common.typeinfo.TypeInformation
+
+/**
+  * A Schema represents a Table's structure
+  */
+class Schema(
+  val columnNames: Array[String],
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/Schema.scala]: private?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/Schema.scala [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+ * limitations under the License.
+ */
+package org.apache.flink.api.table
+
+import org.apache.flink.api.common.typeinfo.TypeInformation
+
+/**
+  * A Schema represents a Table's structure
+  */
+class Schema(
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/Schema.scala]: @fhueske are you fine with the name? Or maybe `TableSchema`?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/Schema.scala [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  val columnNameToIndex: Map[String, Int] = columnNames.zipWithIndex.toMap
+
+  /**
+    * Returns all type informations as an array
+    */
+  def getTypes: Array[TypeInformation[_]] = columnTypes
+
+  /**
+    * Returns the specified type information for the given column index
+ *
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/Schema.scala]: wrong indention

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/Schema.scala [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    * @param columnIndex the index of the field
+    */
+  def getType(columnIndex: Int): TypeInformation[_] = columnTypes(columnIndex)
+
+  /**
+    * Returns the specified type information for the given column name
+ *
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/Schema.scala]: wrong indention

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/Schema.scala [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    * @param columnName the name of the field
+    */
+  def getType(columnName: String): TypeInformation[_] = {
+    if (columnNameToIndex.contains(columnName)) {
+      columnTypes(columnNameToIndex(columnName))
+    } else {
+      throw FieldNotFoundException(s"Table doesn't have the column : $columnName")
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/Schema.scala]: You could also return an `Option`. Then we don't need a new exception type.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/Schema.scala [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    * Returns all type informations as an array
+    */
+  def getTypes: Array[TypeInformation[_]] = columnTypes
+
+  /**
+    * Returns the specified type information for the given column index
+ *
+    * @param columnIndex the index of the field
+    */
+  def getType(columnIndex: Int): TypeInformation[_] = columnTypes(columnIndex)
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/Schema.scala]: No index check here.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/Schema.scala [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    * Returns all column names as an array.
+    */
+  def getColumnNames: Array[String] = columnNames
+
+  /**
+    * Return the specified column name for the given column index.
+ *
+    * @param columnIndex the index of the field
+    */
+  def getColumnName(columnIndex: Int): String = columnNames(columnIndex)
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/Schema.scala]: No index check here. I would use a `Option` here.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala [hunk 8]
```diff
@@ -66,6 +66,17 @@ class Table(
   def getRelNode: RelNode = logicalPlan.toRelNode(relBuilder)
 
   /**
+    * Returns the schema of this Table
+    */
+  def schema: Schema = new Schema(logicalPlan.output.map(_.name).toArray,
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala]: I would call this `getSchema` to be consistent.

File: flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/batch/SchemaTest.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+import org.apache.flink.api.table.Schema;
+import org.apache.flink.api.table.TableException;
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+
+public class SchemaTest {
+
+	@Test
+	public void testSchema() {
```
Comment (twalthr) [inline flink-libraries/flink-table/src/test/java/org/apache/flink/api/java/batch/SchemaTest.java]: Could you add a test that tests getting a schema from a existing table, too?

**Comment by wuchong** [comment]
Hi @twalthr , I addressed the comments. Thank you for your suggestion. I add some tests that getting schema from tables in existing ITCases.

**Comment by twalthr** [comment]
Thanks @wuchong. I think we can convert the modified ITCases to TableTestBase tests. I will try it and merge this.

**Pull request closed without merging**